src/Storage/Database/Schema/Types/JsonType.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Bolt\Storage\Database\Schema\Types;
+
+use Bolt\Common\Exception\DumpException;
+use Bolt\Common\Exception\ParseException;
+use Bolt\Common\Json;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types;
+
+/**
+ * Type generating JSON object values.
+ *
+ * DBAL 2.6 deprecated JsonArrayType in favour of JsonType, and bumped the
+ * minimum PHP version to 7.1. As we need to maintain support for PHP 5.5+
+ * this has caused problems with schema change detection.
+ *
+ * @see https://github.com/bolt/bolt/issues/6856
+ *
+ * @internal Provided to bridge DBAL 2.5 & 2.6 support.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ * @author Gawain Lynch <gawain.lynch@gmail.com>
+ */
+final class JsonType extends Types\Type
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform)
+    {
+        return $platform->getJsonTypeDeclarationSQL($fieldDeclaration);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertToDatabaseValue($value, AbstractPlatform $platform)
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        try {
+            $encoded = Json::dump($value);
+        } catch (DumpException $e) {
+            throw Types\ConversionException::conversionFailed($value, 'json');
+        }
+
+        return $encoded;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convertToPHPValue($value, AbstractPlatform $platform)
+    {
+        if ($value === null || $value === '') {
+            return null;
+        }
+
+        if (is_resource($value)) {
+            $value = stream_get_contents($value);
+        }
+
+        try {
+            $val = Json::parse($value, true);
+        } catch (ParseException $e) {
+            throw Types\ConversionException::conversionFailed($value, $this->getName());
+        }
+
+        return $val;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return 'json';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function requiresSQLCommentHint(AbstractPlatform $platform)
+    {
+        return !$platform->hasNativeJsonType();
```

Comment:
I'm pretty sure that this always needs to return true (matching upstream) that way DBAL doesn't try to use JsonArrayType when mapping back to PHP value.

Reply:
OK, I *think* I know part of what is confusing us here. Running it though xdebug on freshly dropped/recreated PostgreSQL 9.6 tables (which definitely has JSON support :wink:) 

### ContentType
```
koalas:
    name: Koalas
    singular_name: Koala
    fields:
        title:
            type: text
            class: large
            group: content
        slug:
            type: slug
            uses: title
        images:
            type: imagelist
        body:
            type: html
            height: 300px
```
### Schema
```
bolt_testing=# \d+ bolt_koalas
                                                              Table "public.bolt_koalas"
     Column     |              Type              |                        Modifiers                         | Storage  | Stats target |  Description   
----------------+--------------------------------+----------------------------------------------------------+----------+--------------+----------------
 id             | integer                        | not null default nextval('bolt_koalas_id_seq'::regclass) | plain    |              | 
 slug           | character varying(128)         | not null                                                 | extended |              | 
 datecreated    | timestamp(0) without time zone | not null                                                 | plain    |              | 
 datechanged    | timestamp(0) without time zone | not null                                                 | plain    |              | 
 datepublish    | timestamp(0) without time zone | default NULL::timestamp without time zone                | plain    |              | 
 datedepublish  | timestamp(0) without time zone | default NULL::timestamp without time zone                | plain    |              | 
 username       | character varying(32)          | default NULL::character varying                          | extended |              | 
 ownerid        | integer                        |                                                          | plain    |              | 
 status         | character varying(32)          | not null                                                 | extended |              | 
 templatefields | json                           |                                                          | extended |              | (DC2Type:json)
 title          | character varying(256)         | default NULL::character varying                          | extended |              | 
 images         | json                           |                                                          | extended |              | (DC2Type:json)
 body           | text                           |                                                          | extended |              | 
Indexes:
    "bolt_koalas_pkey" PRIMARY KEY, btree (id)
    "idx_aef722b7b00651c" btree (status)
    "idx_aef722b989d9b62" btree (slug)
    "idx_aef722ba5131421" btree (datepublish)
    "idx_aef722bafba6fd8" btree (datecreated)
    "idx_aef722bb7805520" btree (datedepublish)
    "idx_aef722bbe74e59a" btree (datechanged)
```

### Breakpoints
![screenshot from 2017-08-31 06-46-27](https://user-images.githubusercontent.com/1427081/29906847-25ef8490-8e18-11e7-851c-30468e6d1c54.png)

### Record
![screenshot from 2017-08-31 06-48-47](https://user-images.githubusercontent.com/1427081/29906907-8c795970-8e18-11e7-9ad3-a6d26173be60.png)

### Trace
![image](https://user-images.githubusercontent.com/1427081/29906932-bd603c98-8e18-11e7-9758-8a42abef0ae5.png)

### Overridden JSON class
![image](https://user-images.githubusercontent.com/1427081/29906961-f51ac7a2-8e18-11e7-90de-de249259c188.png)

Note the break point is still on the deprecated DBAL class.

Reply:
It is worth noting that not **all** of the JSON fields go to the legacy class (even when repeated with the hinting off), it seems to be just some of the CT ones. I am guessing this is where we are crossing wires?

Reply:
:rofl: I know why … commit(s) incoming

Reply:
OK, so now with the "storage type mapping" commit, I can now start with a fresh DB on above set up, create first user, and create content without hitting the legacy method … Including starting on a fresh 3.2 install.

Reply:
Dear Internet historians … OK, so here is how this turned out.

So in February 2011 Benjamin [added this commit](https://github.com/doctrine/dbal/commit/28c7e09ddec068f0d3abc759858dcba8e05c1067) that enabled Dcotrine types to have comments to determine "Is it necessary for the platform to add a parsable type comment to allow reverse engineering the given type?" via the `$this->doctrineTypeComments` class property.

Now the only places this is used are 
- `AbstractPlatform::isCommentedDoctrineType()`
- `AbstractPlatform::markDoctrineTypeCommented()`
- `AbstractPlatformgetDoctrineTypeComment()` 

… and the only place any of those is called in the Doctrine libraries Bolt includes is in `DB2Platform` 

… and while I am still a closet fan of DB2, we don't use or support it.

But wait, there's more!

Fast forward to 2015 and https://github.com/doctrine/dbal/pull/891 introduced the first attempt to enhance and fix some of the original design shortcomings in `JsonArrayType`, which was correctly pushed back on due to the BC break it would have created.

The PR author then created an updated PR, https://github.com/doctrine/dbal/pull/892, to introduce a new type … hence where we are right here.

In https://github.com/doctrine/dbal/pull/892 there is [this key comment from Christophe](https://github.com/doctrine/dbal/pull/892#r36311026):

> But the native type can only be mapped to a single doctrine type, and all other doctrine types reusing the same native type must use a comment hint to make the schema retrieval work properly.
> Otherwise the native field will always be considered as a json_array and any of your json field will trigger an ALTER TABLE query in each schema update run or migration generation. this is precisely why we introduced comment hints

In other words, as I read it, the comment(s) are only for migrations … and we don't use, or support, migrations.